### PR TITLE
[#164721733] Add back button to calendar modal and change preferences calendar screen

### PR DESCRIPTION
**Pull request by lmorelli986**
This PR aims to:
* enable back button feature to calendar selection modal in messageCTAs
* change preferences calendar screen from modal to normal navigation screen

calendar

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🐞 [#164721733](https://www.pivotaltracker.com/story/show/164721733): Quando apro la modale per la scelta del calendario non é possolibile usare il tasto back fisico

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: ts/screens/preferences/CalendarScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+ * scenarios:
+ * * a switch to enable/disable biometric recognition. Enabled if the
+ *   fingerprint is enrolled.
+ * * the above switch (disabled) with a button below that redirects the user to
+ *   device settings otherwise.
+ *
+ * Please note that this screen must not be displayed if biometric recognition
+ * is unavailable.
+ */
+/**
```
Comment (cloudify) [inline ts/screens/preferences/CalendarScreen.tsx]: is this comment about this class? (then it should be merged with the comment above) or is it misplaced?

File: ts/screens/preferences/CalendarScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+ * A modal that allow the user to select one of the device available Calendars
+ */
+class CalendarScreen extends React.PureComponent<Props, State> {
```
Comment (cloudify) [inline ts/screens/preferences/CalendarScreen.tsx]: I would state that this is about preferences, `CalendarScreen` seems like a screen that shows a calendar - perhaps `CalendarsPreferencesScreen` ? (also route name should be aligned)

File: ts/screens/preferences/CalendarScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+  constructor(props: Props) {
+    super(props);
+    this.state = INITIAL_STATE;
+  }
+
+  private goBack = () => this.props.navigation.goBack();
```
Comment (cloudify) [inline ts/screens/preferences/CalendarScreen.tsx]: why not just use `this.props.navigation.goBack` below ?

File: ts/screens/preferences/CalendarScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+              ))}
+            </React.Fragment>
+          )}
+          <Text style={styles.separator}>{I18n.t("messages.cta.helper")}</Text>
+        </View>
+      </TopScreenComponent>
+    );
+  }
+
+  public componentDidMount() {
```
Comment (cloudify) [inline ts/screens/preferences/CalendarScreen.tsx]: move this at top, before the render, so it's clear what happens on mount (before render)

File: ts/screens/preferences/CalendarScreen.tsx [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    this.fetchCalendars();
+  }
+
+  private fetchCalendars = () => {
+    this.setState({ calendars: pot.toLoading(pot.none) });
```
Comment (cloudify) [inline ts/screens/preferences/CalendarScreen.tsx]:
```suggestion
    this.setState({ calendars: pot.noneLoading });
```

`toLoading` can be used to convert a previous pot value to loading

File: ts/screens/preferences/CalendarScreen.tsx [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    // Fetch user calendars.
+    // The needed permissions are already checked/asked by the MessageCTABar component.
+    RNCalendarEvents.findCalendars()
+      .then(calendars =>
+        // Filter only the calendars that allow modifications
+        calendars.filter(calendar => calendar.allowsModifications)
+      )
+      .then(calendars => this.setState({ calendars: pot.some(calendars) }))
```
Comment (cloudify) [inline ts/screens/preferences/CalendarScreen.tsx]: the two `then` blocks can be safely merged, I guess, as `filter` doesn't throw an exception

File: ts/screens/preferences/PreferencesScreen.tsx [hunk 7]
```diff
@@ -46,33 +41,21 @@ type OwnProps = Readonly<{
...
 type Props = OwnProps &
-  LightModalContextInterface &
   ReturnType<typeof mapStateToProps> &
   ReturnType<typeof mapDispatchToProps> &
   ReduxProps;
 
 type State = {
   isFingerprintAvailable: boolean;
   hasCalendarPermission: boolean;
+  checkCalendarPermissionAndUpdateStateListener?: any;
```
Comment (cloudify) [inline ts/screens/preferences/PreferencesScreen.tsx]: no better type than `any` ? last resort:

```ts
ReturnType<typeof this.props.navigation.addListener>
```

File: ts/screens/preferences/PreferencesScreen.tsx [hunk 8]
```diff
@@ -108,35 +91,31 @@ class PreferencesScreen extends React.Component<Props, State> {
...
+      checkCalendarPermissionAndUpdateStateListener: this.props.navigation.addListener(
+        "willFocus",
+        this.checkCalendarPermissionAndUpdateState
+      )
+    });
+  }
+
+  public componentWillUnmount() {
+    if (
+      typeof this.state.checkCalendarPermissionAndUpdateStateListener !==
```
Comment (cloudify) [inline ts/screens/preferences/PreferencesScreen.tsx]:
```suggestion
      this.state.checkCalendarPermissionAndUpdateStateListener !== undefined
```

File: ts/screens/preferences/PreferencesScreen.tsx [hunk 9]
```diff
@@ -205,7 +184,9 @@ class PreferencesScreen extends React.Component<Props, State> {
               </ListItem>
             )}
             {hasCalendarPermission && (
-              <ListItem onPress={this.renderDefaultCalendarPreference}>
+              <ListItem
+                onPress={() => this.props.navigateToCalendarPreferenceScreen()}
```
Comment (cloudify) [inline ts/screens/preferences/PreferencesScreen.tsx]:
```suggestion
                onPress={this.props.navigateToCalendarPreferenceScreen}
```

File: ts/screens/preferences/CalendarScreen.tsx (hunk 4 above)
Comment (fpersico) [reply ts/screens/preferences/CalendarScreen.tsx]: Just to be clear componentDidMount is called AFTER render

File: ts/components/SelectCalendarModal.tsx [hunk 10]
```diff
@@ -99,6 +100,11 @@ class SelectCalendarModal extends React.PureComponent<Props, State> {
     };
   }
 
+  private onBackPress = () => {
+    this.props.onCancel();
+    return true;
```
[the remark this answers is not in the document]
Comment (lmorelli986) [reply ts/components/SelectCalendarModal.tsx]: comment added again

File: ts/screens/preferences/CalendarScreen.tsx (hunk 2 above)
Comment (lmorelli986) [reply ts/screens/preferences/CalendarScreen.tsx]: moved to CalendarsPreferencesScreen

File: ts/screens/preferences/CalendarScreen.tsx (hunk 5 above)
Comment (lmorelli986) [reply ts/screens/preferences/CalendarScreen.tsx]: changed

File: ts/screens/preferences/CalendarScreen.tsx (hunk 6 above)
Comment (lmorelli986) [reply ts/screens/preferences/CalendarScreen.tsx]: merged the two blocks

File: ts/screens/preferences/PreferencesScreen.tsx (hunk 7 above)
Comment (lmorelli986) [reply ts/screens/preferences/PreferencesScreen.tsx]: unfortunately in this point `this` is out of class scope, I had to add 
```
ReturnType<
    NavigationScreenProp<NavigationState>["addListener"]
  >;
```

File: ts/screens/preferences/PreferencesScreen.tsx (hunk 8 above)
Comment (lmorelli986) [reply ts/screens/preferences/PreferencesScreen.tsx]: changed

File: ts/screens/preferences/PreferencesScreen.tsx (hunk 9 above)
Comment (lmorelli986) [reply ts/screens/preferences/PreferencesScreen.tsx]: changed

File: ts/screens/preferences/CalendarScreen.tsx (hunk 3 above)
Comment (lmorelli986) [reply ts/screens/preferences/CalendarScreen.tsx]: changed

File: ts/screens/preferences/CalendarScreen.tsx (hunk 4 above)
Comment (lmorelli986) [reply ts/screens/preferences/CalendarScreen.tsx]: would you prefer to move it above `render()` anyway? @cloudify

Comment (cloudify) [reply ts/screens/preferences/CalendarScreen.tsx]: it's ok as it is, thanks

**Pull request merged**